Compute bounds for the variables created during codegen

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123100

Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

lezcano committed:
Compute bounds for the variables created during codegen

Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/common.py:
```diff
@@ -1455,31 +1460,62 @@ def __enter__(self):
+                    # We will also likely not get much from computing VRs on these nodes
+                    if any(
+                        s in fx_node.target
+                        for s in ("set_indirect", "reduction", "scan")
```
Not necessary for now, but some reductions and/or scans should be really easy. e.g. `min`/`max` doesn't change the bounds at all.

lezcano replied:
We should probably fix that in the pass that computes the bounds in the FX graph, not here, but yep.

peterbell10 commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -632,6 +635,11 @@ def where(a, b, c):
+        if a == ValueRanges.unknown():
+            return b
+        if b == ValueRanges.unknown():
+            return a
```
This makes no sense? Isn't `a` a completely different dtype from `b` and `c`.

lezcano replied:
Ugh, right, this doesn't make sense at all...

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -850,11 +850,9 @@ def floordiv(a, b):
+        left = ops.to_dtype((ops.lt(z, x)), torch.int8)
+        right = ops.to_dtype((ops.lt(x, z)), torch.int8)
         sub = ops.sub(left, right)
         return f"{sub}.to({x}.dtype)"
```
What's the issue that requires all these to be updated. I'm not opposed, just curious.

lezcano replied:
With this approach, either all the lowerings are written with strings or all of them are written with ops.If you mix them, you end up getting strings as the inputs in the bound propagation and we don't know how to "bound these strings". You can sidestep this limitation using CSE variables and setting the bounds properly on them, but yeah.

lezcano replied:
Let me write a comment in the part where we forward the bounds.

peterbell10 commented on torch/_inductor/codegen/common.py:
```diff
@@ -569,7 +570,11 @@ def bitwise_right_shift(x, y):
-        return ops.where(f"(({r} != 0) & (({r} < 0) != ({b} < 0)))", ops.add(r, b), r)
+        cond = ops.and_(
+            ops.ne(r, ops.constant(0, torch.int32)),
+            ops.ne(ops.signbit(r), ops.signbit(b)),
```
Why change to signbit here?

lezcano replied:
It's just more descriptive of what's going on here.

peterbell10 replied:
It's also a much more complicated compound op instead of a divice primitive though.

lezcano replied:
We do lower it to `x < 0` in triton for integer dtypes and to signbit otherwise in triton and C++. I figured that it'd be pretty much equivalent to an `&` by a constant, and I figured it'd be equivalent to `x < 0`.

Anyway, not a big deal, reverted.

peterbell10 replied:
Oh, I was mixing this up with `ops.sign` . This is lowered as `libdevice.signbit` which I would expect to have a 1 cycle latency, so yeah not an issue.

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -892,10 +892,20 @@ def constant(val, dtype):
 
     @staticmethod
     def index_expr(expr, dtype):
+        bounds = ValueRanges.unknown()
+        # If this expression does not come from an FX node, we compute its bounds
+        if (
+            fx_node := getattr(V.interpreter, "current_node", None)
+        ) and fx_node.target != "index_expr":
+            bounds = bound_sympy(expr)
```
Even if it comes from the fx node, we should compute the bounds with `bound_sympy(expr)` anyway, right?

lezcano replied:
When it comes from the fx_node, we'll forward the already-computed nodes in the `getattr` in CSEProxy

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1481,7 +1491,15 @@ def index_expr(expr, dtype):
             return CppOverrides.index_expr(expr, dtype)
         elif stride is not None:
-            value = ops.to_dtype(cexpr(index), dtype)
+            bounds = ValueRanges.unknown()
+            # If this expression does not come from an FX node, we compute its bounds
+            if (
+                fx_node := getattr(V.interpreter, "current_node", None)
+            ) and fx_node.target != "index_expr":
+                bounds = bound_sympy(index)
+
+            idx = V.kernel.cse.generate(V.kernel.compute, cexpr(index), bounds=bounds)
```
nit: perhaps we can factor out a util function for this, same code patterns in several places.

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented:
This one is ready for review.
I have put the logic in this PR under a flag, which I will turn off by default, as I don't think that running this analysis is beneficial at the moment (we don't use the bounds for intermediary variables anywhere).

That being said, I think that it is still worth it to land this PR, as it shows how to propagate a property from the FX graph to pretty much every CSE variable that is created via `ops.`. This should be used by the C++ codegen to type all the variables properly (at the moment the C++ backend does a very approximate inference of the dtypes of variables) @jgong5

peterbell10 reviewed (commented):
Is it possible to test this?

lezcano commented:
I thought about it, but I didn't come up with any reasonable way.
I did check manually tho, and everything looks fine.

Note that most of the heavy lifting is done by the methods we already have to compute bounds.

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano commented:
It will be much easier to test this when dtype propagation is implemented, as we then can statically_assert in C++ that our inference was indeed correct

lezcano commented:
Actually, this can be tested by putting up a test flag that prints every bound into a device_assert. Now, God knows that would fail left right and centre, so I'm leaving it as it is for now. I might use this task as an onboarding task tho.

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting you change but it is failing inductor tests https://hud.pytorch.org/pytorch/pytorch/commit/bb668c6468dd4adf7737a069e7af4c3f612cfc81' -c landrace

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "Compute bounds for the variables created during codegen"


Before we would just bail out on these bounds for all variables that did
not come from the FX graph. Now we propagate the bounds whenever we have
a rule for that op.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 gujinghui PenghuiCheng jianyuh min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge
